components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java

```diff
@@ -0,0 +1,223 @@
+                log.debug("Authorized user: " + authorizedUser.toFullQualifiedUsername() + " not associated to: " +
+                        associatedUser.toFullQualifiedUsername());
+            }
+            ResponseHeader responseHeader = new ResponseHeader();
+            responseHeader.setKey("error-description");
+            responseHeader.setValue("Associated user is invalid.");
+            tokReqMsgCtx.addProperty("RESPONSE_HEADERS", new ResponseHeader[]{responseHeader});
+
+            return false;
+        }
+
+        tokReqMsgCtx.setAuthorizedUser(
+                OAuth2Util.getUserFromUserName(associatedUser.toFullQualifiedUsername()));
+
+        String[] allowedScopes =  getAllowedScopes(validationResponseDTO.getScope(),
+                tokReqMsgCtx.getOauth2AccessTokenReqDTO().getScope());
+        tokReqMsgCtx.setScope(allowedScopes);
+
+        if (log.isDebugEnabled()) {
+            log.debug("Issuing an access token for associated user: " + associatedUser + " with scopes: " +
+                    tokReqMsgCtx.getScope());
+        }
+
+        return true;
+    }
+
+    private String extractParameter(String param, OAuthTokenReqMessageContext tokReqMsgCtx) {
+
+        RequestParameter[] parameters = tokReqMsgCtx.getOauth2AccessTokenReqDTO().getRequestParameters();
+
+        for (RequestParameter parameter : parameters) {
+            if (param.equals(parameter.getKey())) {
+                if (parameter.getValue() != null && parameter.getValue().length > 0) {
+                    return parameter.getValue()[0];
+                }
+            }
+        }
+
+        return null;
+    }
+
+    private String[] getAllowedScopes(String[] authorizedScoped, String[] requestedScopes) {
+
+        HashSet<String> allowedScopesSet = new HashSet<>();
+        allowedScopesSet.addAll(Arrays.asList(authorizedScoped));
+        allowedScopesSet.retainAll(Arrays.asList(requestedScopes));
+
+        if (log.isDebugEnabled()) {
+            log.debug("Allowed scopes: " + allowedScopesSet);
+        }
+
+        String[] allowedScopes = {};
+        allowedScopes = allowedScopesSet.toArray(allowedScopes);
+        return allowedScopes;
+    }
+
+    private boolean validateAssociation(User user, User associatedUser) throws UserAccountAssociationException {
+
+        UserAccountAssociationDTO[] userAccountAssociationDTOS =
+                UserAccountConnectorImpl.getInstance().getAccountAssociationsOfUser(user.toFullQualifiedUsername());
+        if (userAccountAssociationDTOS != null) {
+            for (UserAccountAssociationDTO userAccountAssociationDTO : userAccountAssociationDTOS) {
+
+                if (userAccountAssociationDTO.getTenantDomain().equals(associatedUser.getTenantDomain())) {
+                    if (userAccountAssociationDTO.getDomain().equals(associatedUser.getUserStoreDomain())) {
+                        if (userAccountAssociationDTO.getUsername().equals(associatedUser.getUserName())) {
```

Comment:
Do we need to consider user name case sensitiveness here?

Reply:
No need due to the same mentioned in https://github.com/wso2-extensions/identity-user-account-association/pull/30/files#r339900959